### PR TITLE
UCT/DEVX/SIG: Add signature debug

### PR DESCRIPTION
**Pull request by Artemy-Mellanox**

### CODE REVIEW

File: src/uct/ib/mlx5/dv/sig.c [hunk 1]
```diff
@@ -218,6 +220,18 @@ static void* uct_ib_mlx5_sig_set_stride(void* seg, uint32_t lkey, size_t size,
     return UCS_PTR_TYPE_OFFSET(seg, (*sentry));
 }
 
+static void* uct_ib_mlx5_set_klm(void* seg, uint32_t lkey, size_t size,
+                                 void* buff)
+{
+    struct mlx5_wqe_umr_klm_seg *klm = seg;
+
+    klm->byte_count = htonl(size);
```
Comment (gleon99) [inline src/uct/ib/mlx5/dv/sig.c:232]: Align `=`

File: src/uct/ib/mlx5/dv/sig.c [hunk 2]
```diff
@@ -218,6 +220,18 @@ static void* uct_ib_mlx5_sig_set_stride(void* seg, uint32_t lkey, size_t size,
     return UCS_PTR_TYPE_OFFSET(seg, (*sentry));
 }
 
+static void* uct_ib_mlx5_set_klm(void* seg, uint32_t lkey, size_t size,
```
Comment (gleon99) [inline src/uct/ib/mlx5/dv/sig.c:227]: `const` where possible

File: src/uct/ib/mlx5/dv/sig.c [hunk 3]
```diff
@@ -19,12 +19,14 @@
...
 #define UCT_IB_MLX5_DIF_IPCS                          0x2
-#define UCT_IB_MLX5_BLOCK_SIZE_IDX                    0x6
+#define UCT_IB_MLX5_BLOCK_SIZE_IDX_512                0x1
+#define UCT_IB_MLX5_BLOCK_SIZE_IDX_4048               0x6
 #define UCT_IB_MLX5_STRIDE_BLOCK_OP                   0x400
 #define UCT_IB_MLX5_BSF_INL_VALID                     UCS_BIT(15)
 #define UCT_IB_MLX5_BSF_REPEAT_BLOCK                  UCS_BIT(7)
 #define UCT_IB_MLX5_MKEY_BSF_EN                       UCS_BIT(30)
 #define UCT_IB_MLX5_WQE_UMR_CTRL_MKEY_MASK_BSF_ENABLE UCS_BIT(12)
+#define UCS_IB_MLX5_SIGNATURE_BLOCK_512               512
```
Comment (gleon99) [inline src/uct/ib/mlx5/dv/sig.c]: IMO such defines (`#define ....512 512`) are redundant&weird.
WDYT of:
1. Rename to `SMALL/LARGE` block
2. Have some array/enum defining those sizes [512, 4048, ] instead?

File: src/uct/ib/base/ib_md.c [hunk 4]
```diff
@@ -191,6 +197,13 @@ static ucs_config_field_t uct_ib_md_config_table[] = {
      "Maximal number of invalidated memory keys that are kept idle before reuse.",
      ucs_offsetof(uct_ib_md_config_t, ext.max_idle_rkey_count), UCS_CONFIG_TYPE_UINT},
 
+#if ENABLE_DEBUG_DATA
+    {"SIG_DEBUG_MODE", "full",
```
Comment (gleon99) [inline src/uct/ib/base/ib_md.c:201]: Does that mean `sig_mode` doesn't get a default value if !DEBUG?
IMO having a struct member which always exists but gets initialized only on conditional-compilation is not so good. Even though currently it doesn't cause a problem (as the logic in `uct_ib_mlx5_sig_mr_init` seems robust enough to handle it), that might not be the case in future.

File: examples/ucp_client_server.c [hunk 5]
```diff
@@ -737,6 +737,7 @@ static int parse_cmd(int argc, char *const argv[], char **server_addr,
             break;
         case 'S':
             sig = 1;
+            sig_block = atoi(optarg);
```
Comment (gleon99) [inline examples/ucp_client_server.c:742]: Need to fix usage string

File: src/uct/ib/base/ib_md.c [hunk 6]
```diff
@@ -191,6 +197,13 @@ static ucs_config_field_t uct_ib_md_config_table[] = {
...
+     "When calculating signature enable partial payload processing",
```
Comment (gleon99) [inline src/uct/ib/base/ib_md.c]: `Enable X when Y`

File: src/uct/ib/mlx5/dv/sig.c [hunk 7]
```diff
@@ -266,8 +280,13 @@ ucs_status_t uct_ib_mlx5_sig_mr_init(uct_ib_mlx5_md_t *md,
     unsigned num_elems;
     size_t length;
     void *wqe;
+    int bs;
 
-    if (sig_attr->block != UCS_IB_MLX5_SIGNATURE_BLOCK_4048) {
+    if (sig_attr->block == UCS_IB_MLX5_SIGNATURE_BLOCK_4048) {
```
Comment (gleon99) [inline src/uct/ib/mlx5/dv/sig.c]: See comment on `#define`s - then this code can be simplified (e.g. access by index)

File: src/uct/ib/mlx5/dv/sig.c [hunk 8]
```diff
@@ -304,15 +323,36 @@ ucs_status_t uct_ib_mlx5_sig_mr_init(uct_ib_mlx5_md_t *md,
...
+        wqe = uct_ib_mlx5_txwq_wrap_exact(txwq, wqe);
+    } else if (md->super.config.sig_mode == UCT_IB_SIG_KLM) {
+        wqe = uct_ib_mlx5_sig_umr_start(txwq, 4, 0, sig->block * num_elems);
+        wqe = uct_ib_mlx5_txwq_wrap_exact(txwq, wqe);
+        wqe = uct_ib_mlx5_set_klm(wqe, sig->mr->lkey, sig->block * num_elems,
+                                  sig->start);
+        wqe = uct_ib_mlx5_sig_umr_round_bb(wqe);
+        wqe = uct_ib_mlx5_txwq_wrap_exact(txwq, wqe);
+    } else {
+        ucs_error("Wrong signature debug mode");
```
Comment (gleon99) [inline src/uct/ib/mlx5/dv/sig.c]: Print the mode

File: src/uct/ib/mlx5/dv/sig.c [hunk 9]
```diff
@@ -282,11 +286,18 @@ ucs_status_t uct_ib_mlx5_sig_mr_init(uct_ib_mlx5_md_t *md,
...
+    }
+
+    for (bs = 0; bs < ucs_static_array_size(uct_ib_mlx5_sig_block); bs++) {
+        if (uct_ib_mlx5_sig_block[bs] == sig_attr->block) {
+            break;
+        }
+    }
+
+    if (bs == ucs_static_array_size(uct_ib_mlx5_sig_block)) {
         ucs_error("unsuported signature block size: %zd", sig_attr->block);
```
Comment (gleon99) [inline src/uct/ib/mlx5/dv/sig.c]: `unsupported`

File: src/uct/ib/mlx5/dv/sig.c [hunk 10]
```diff
@@ -282,11 +286,18 @@ ucs_status_t uct_ib_mlx5_sig_mr_init(uct_ib_mlx5_md_t *md,
     void *wqe;
     int bs;
 
-    if (sig_attr->block == UCS_IB_MLX5_SIGNATURE_BLOCK_4048) {
-        bs = UCT_IB_MLX5_BLOCK_SIZE_IDX_4048;
-    } else if (sig_attr->block == UCS_IB_MLX5_SIGNATURE_BLOCK_512) {
-        bs = UCT_IB_MLX5_BLOCK_SIZE_IDX_512;
-    } else {
+    if (sig_attr->stride < sig_attr->block) {
+        ucs_error("signature stide should be greater or equal to block size");
```
Comment (gleon99) [inline src/uct/ib/mlx5/dv/sig.c]: print both values?

File: src/uct/ib/base/ib_md.c (hunk 4 above)
Comment (Artemy-Mellanox) [reply src/uct/ib/base/ib_md.c:201]: added initialization
https://github.com/Mellanox/ucx/pull/42/commits/239e23e8f160c30bb57e9db665492c4c9f12d6cf#diff-975534026ea952036f6617e7a106314af680d3ca405d777ab04e0af2b427499aR1641

**Comment by gleon99** [comment]
@Artemy-Mellanox please squash.

**Pull request merged**